lib/json/sail_config.c

```diff
@@ -0,0 +1,201 @@
+  cJSON_Hooks hooks;
+  hooks.malloc_fn = &sail_malloc;
+  hooks.free_fn = &sail_free;
+  cJSON_InitHooks(&hooks);
+
+  FILE *f = fopen(path, "rb");
+  fseek(f, 0, SEEK_END);
+  long fsize = ftell(f);
+  fseek(f, 0, SEEK_SET);
+
+  char *buffer = (char *)sail_malloc(fsize + 1);
+  fread(buffer, fsize, 1, f);
+  buffer[fsize] = 0;
+  fclose(f);
+
+  sail_config = cJSON_Parse(buffer);
+
+  if (!sail_config) {
+    sail_assert(false, "Failed to parse configuration");
+  }
+
+  sail_free(buffer);
+}
+
+void sail_config_cleanup(void)
+{
+  sail_free(sail_config);
+}
+
+void sail_config_get_string(sail_string *str, size_t n, const char *key[])
+{
+  cJSON *json = (cJSON *)sail_config;
+
+  sail_free(*str);
+
+  for (int i = 0; i < n; i++) {
+    if (cJSON_IsObject(json)) {
+      json = cJSON_GetObjectItemCaseSensitive(json, key[i]);
+    } else {
+      sail_assert(false, "Failed to access config item");
+    }
+  }
+
+  if (cJSON_IsString(json)) {
+    *str = cJSON_GetStringValue(json);
+  } else {
+    sail_assert(false, "Expected string value in config");
+  }
+}
+
+sail_config_json sail_config_get(size_t n, const char *key[])
+{
+  sail_config_json result;
+  cJSON *json = (cJSON *)sail_config;
+
+  for (int i = 0; i < n; i++) {
+    if (cJSON_IsObject(json)) {
+      json = cJSON_GetObjectItemCaseSensitive(json, key[i]);
+    } else {
+      sail_assert(false, "Failed to access config item");
+    }
+  }
+
+  return (sail_config_json)json;
+}
+
+bool sail_config_is_object(const sail_config_json config)
+{
+  return cJSON_IsObject((cJSON *)config);
+}
+
+bool sail_config_object_has_key(const sail_config_json config, const sail_string key)
+{
+  return cJSON_HasObjectItem((cJSON *)config, key);
+}
+
+sail_config_json sail_config_object_key(const sail_config_json config, const sail_string key)
+{
+  return (sail_config_json)cJSON_GetObjectItemCaseSensitive((cJSON *)config, key);
+}
+
+bool sail_config_is_string(const sail_config_json config)
+{
+  return cJSON_IsString((cJSON *)config);
+}
+
+bool sail_config_is_array(const sail_config_json config)
+{
+  return cJSON_IsArray((cJSON *)config);
+}
+
+bool sail_config_is_bool_array(const sail_config_json config)
+{
+  if (!sail_config_is_array(config)) {
+    return false;
+  }
+
+  int len = cJSON_GetArraySize((cJSON *)config);
+
+  cJSON *value;
+  cJSON_ArrayForEach(value, ((cJSON*)config)) {
+    if (!cJSON_IsBool(value)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool sail_config_is_bool_array_with_size(const sail_config_json config, mach_int expected)
+{
+  if (!sail_config_is_bool_array(config)) {
+    return false;
+  }
+
+  int len = cJSON_GetArraySize((cJSON *)config);
+
+  return (mach_int)len == expected;
+}
+
+void sail_config_unwrap_string(sail_string *str, const sail_config_json config)
+{
+  *str = cJSON_GetStringValue((cJSON *)config);
+}
+
+void sail_config_unwrap_int(sail_int *n, const sail_config_json config)
+{
+  char *str = cJSON_GetStringValue((cJSON *)config);
```

Comment:
As I understand it this means the JSON never uses JSON numbers - only strings of numbers?

I guess that makes sense, especially because `cjson`'s number handling looks whack (silently clamping to INT_MAX???), but it's a slight shame. The JSON spec doesn't actually put a limit on integer size so we theoretically could support arbitrary precision integers "natively". Although I guess that might make it hard to generate them from some programming languages.... And it would require using a better parser. Hmmm.

Reply:
Yes, although I probably do need to use integers for the JSON schema output to work correctly. So maybe I will need to find a better parser.